### PR TITLE
DPDK: Fix compilation.

### PR DESCRIPTION
**Pull request by AndersBroman**
I'm not sure this works as it should. I have installed DPDK on my system but are struggling to configure it properly. So in this state when doing:
/build-pcap$ dumpcap -D
USER1: permission denied, DPDK needs root permission
dumpcap: Can't get list of interfaces: 

Isn't this supposed to print the other interfaces?
This also fais:
uild-pcap$ sudo dumpcap -D
EAL: Cannot obtain physical addresses: No such file or directory. Only vfio will function.
error allocating rte services array
EAL: FATAL: rte_service_init() failed
EAL: rte_service_init() failed
dumpcap: Can't get list of interfaces:

### CODE REVIEW

**Comment by AndersBroman** [comment]
Line 980 pcap_snprintf(dpdk_pre_init_errbuf, PCAP_ERRBUF_SIZE,
Should that be:
pcap_snprintf(ebuf, PCAP_ERRBUF_SIZE,

If I change line 983 ret = PCAP_ERROR;
to 
ret = PCAP_ERROR_NO_SUCH_DEVICE;

I get:
$ dumpcap -D
USER1: permission denied, DPDK needs root permission
1. eth0
2. eth3
3. lo (Loopback)
4. any
5. eth1
6. nflog
7. nfqueue
8. dbus-system
9. dbus-session

sudo dumpcap -D
EAL: Cannot obtain physical addresses: No such file or directory. Only vfio will function.
error allocating rte services array
EAL: FATAL: rte_service_init() failed
EAL: rte_service_init() failed
1. eth0
2. eth3
3. lo (Loopback)
4. any
5. eth1
6. nflog
7. nfqueue
8. dbus-system
9. dbus-session

Without the 933 change  I get:

umpcap -D
USER1: permission denied, DPDK needs root permission
1. eth0
2. eth3
3. lo (Loopback)
4. any
5. eth1
6. nflog
7. nfqueue
8. dbus-system
9. dbus-session
TCL-sniffer3:~/build-pcap$ sudo dumpcap -D
EAL: Cannot obtain physical addresses: No such file or directory. Only vfio will function.
error allocating rte services array
EAL: FATAL: rte_service_init() failed
EAL: rte_service_init() failed
dumpcap: Can't get list of interfaces: Can't open DPDK device: A service core failed to launch successfully

**Comment by guyharris** [comment]
What happens if you just remove the `geteuid()` check from `dpdk_pre_init()`?

File: pcap-dpdk.c [hunk 1]
```diff
@@ -710,7 +710,8 @@ static int dpdk_pre_init(char * ebuf)
 			// as an error.
 			//
 			dpdk_fmt_errmsg_for_rte_errno(ebuf,
-			    PCAP_ERRBUF_SIZE, -is_dpdk_pre_inited);
+			    PCAP_ERRBUF_SIZE, -is_dpdk_pre_inited,
```
Comment (guyharris) [inline pcap-dpdk.c]: This one's fixed in 2ed0710c942d5c8ed166bafee1cea7195ab791e5.

File: pcap-dpdk.c [hunk 2]
```diff
@@ -977,10 +978,10 @@ int pcap_dpdk_findalldevs(pcap_if_list_t *devlistp, char *ebuf)
 		if (ret < 0)
 		{
 			// This returns a negative value on an error.
-			pcap_snprintf(p->errbuf, PCAP_ERRBUF_SIZE,
-			    "Can't open device %s: %s",
-			    p->opt.device, dpdk_pre_init_errbuf);
-			ret = PCAP_ERRNO;
+			pcap_snprintf(dpdk_pre_init_errbuf, PCAP_ERRBUF_SIZE,
```
Comment (guyharris) [inline pcap-dpdk.c]: This one should be fixed in 92f46849317edf71217dca872e516feeffb364cc.

**Comment by AndersBroman** [comment]
Hi,
It travis can be made to use Debian DPDK can be found here
https://packages.debian.org/search?keywords=dpdk
>

What happens if you just remove the geteuid() check from dpdk_pre_init()?
I'll try that when I have access to my remote server.
Regards
Anders Broman

Den lör 9 feb. 2019 kl 01:04 skrev Guy Harris <notifications@github.com>:

**Comment by guyharris** [comment]
You might want to rebase this change and resolve the conflicts.

**Comment by guyharris** [comment]
> Hi, It travis can be made to use Debian DPDK can be found here https://packages.debian.org/search?keywords=dpdk

[It appears that you can install APT packages from a custom repository](https://docs.travis-ci.com/user/installing-dependencies/#installing-packages-from-a-custom-apt-repository).  However, they say

> If you’re installing packages this way, make sure you download the correct version for your environment.

If [this answer](https://askubuntu.com/questions/445487/what-debian-version-are-the-different-ubuntu-versions-based-on) is to be believed, Trusty is based on jessie and Xenial is based on stretch, and t[here's a jessie-backports DPDK package based on DPDK 16.11](https://packages.debian.org/jessie-backports/dpdk), so that *might* be a new enough version of DPDK to support compiling the libpcap DPDK support - [it does, at least, have `rte_delay_us_block()`](http://doc.dpdk.org/api-16.11/rte__cycles_8h.html#ab34819cd1672d3807a9ed341f65e8c15).

I'll see if I can install that version.

**Comment by guyharris** [comment]
> I'll see if I can install that version.

Well, that was entertaining.

At least if you use the APT add-on, you can specify the jessie-backports repository...

...but you appear to need to provide a URL for its GPG key, and I have failed to find anything online that says "here's the URL for that repository's GPG key" or even "here's the URL for the GPG key for all of Debian's repositories".

**Comment by AndersBroman** [comment]
I added the v18.11 packages to my LTS 18.04 Bionic by adding this line:
deb [trusted=yes] http://ftp.se.debian.org/debian/ testing main contrib
to /etc/apt/sourceslist it stil whines about the GPG Key but loads the package(s), Perhaps adding [trusted=yes] would work for travis too

**Comment by AndersBroman** [comment]
With current tip of mastrer i get:
dumpcap -D
USER1: permission denied, DPDK needs root permission
dumpcap: Can't get list of interfaces: Can't look for DPDK devices: DPDK
requires that it run as root
@TCL-sniffer3:~/build-pcap$
@TCL-sniffer3:~/build-pcap$
@TCL-sniffer3:~/build-pcap$ sudo dumpcap -D
EAL: Cannot obtain physical addresses: No such file or directory. Only vfio
will function.
error allocating rte services array
EAL: FATAL: rte_service_init() failed
EAL: rte_service_init() failed
dumpcap: Can't get list of interfaces: Can't look for DPDK devices: A
service core failed to launch successfully

Shouldn't it list the other interfaces?
Reghards
Anders


Den sön 10 feb. 2019 kl 06:03 skrev Guy Harris <notifications@github.com>:

**Comment by AndersBroman** [comment]
Den sön 10 feb. 2019 kl 17:13 skrev Anders Broman <a.broman58@gmail.com>:

> With current tip of mastrer i get:
> dumpcap -D
> USER1: permission denied, DPDK needs root permission
> dumpcap: Can't get list of interfaces: Can't look for DPDK devices: DPDK
> requires that it run as root
> @TCL-sniffer3:~/build-pcap$
> @TCL-sniffer3:~/build-pcap$
> @TCL-sniffer3:~/build-pcap$ sudo dumpcap -D
> EAL: Cannot obtain physical addresses: No such file or directory. Only
> vfio will function.
> error allocating rte services array
> EAL: FATAL: rte_service_init() failed
> EAL: rte_service_init() failed
> dumpcap: Can't get list of interfaces: Can't look for DPDK devices: A
> service core failed to launch successfully
>
> Shouldn't it list the other interfaces?
> Regards
> Anders
>
>
Deleting the check for root permissions yield the following:
@TCL-sniffer3:~/build-pcap$ dumpcap -D
USER1: permission denied, DPDK needs root permission
dumpcap: Can't get list of interfaces: Can't look for DPDK devices: DPDK
requires that it run as root
ericsson@TCL-sniffer3:~/build-pcap$ sudo dumpcap -D
EAL: Cannot obtain physical addresses: No such file or directory. Only vfio
will function.
error allocating rte services array
EAL: FATAL: rte_service_init() failed
EAL: rte_service_init() failed
dumpcap: Can't get list of interfaces: Can't look for DPDK devices: A
service core failed to launch successfully
ericsson@TCL-sniffer3:~/build-pcap$

So it seems like the check is not needed.
Best regards
Anders

**Comment by guyharris** [comment]
> So it seems like the check is not needed.

I removed that check in abeb357041f40ccf2dd715cab5e2860ceb851eb7.

**Comment by guyharris** [comment]
> Shouldn't it list the other interfaces?

I've changed it in d08338fdff4538ea18fa3b94327a909f9ed1de46 so that, for pcap_findalldevs(), an EACCESS when initializing DPDK just causes it not to list DPDK devices rather than failing.

**Comment by guyharris** [comment]
Try running both not as root, and as root, with `DPDK_CFG=--log-level=debug`:

    DPDK_CFG=--log-level=debug dumpcap -D
    DPDK_CFG=--log-level=debug sudo dumpcap -D

**Comment by AndersBroman** [comment]
Hi,
I think I messed up testing without sudo yesterday but I think removing the
code from pacap-dpdk is the right thing
to do any way as I thjink it's possible to set DPDK up as to not require
root privileges.

grep Huge /proc/meminfo
AnonHugePages:         0 kB
ShmemHugePages:        0 kB
HugePages_Total:       2
HugePages_Free:        2
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:    1048576 kB

@TCL-sniffer3:~/build-pcap$ DPDK_CFG=--log-level=debug dumpcap -D
EAL: Detected 12 lcore(s)
EAL: Detected 1 NUMA nodes
EAL: Multi-process socket /run/user/1002/dpdk/rte/mp_socket
EAL: Probing VFIO support...
EAL: rte_mem_virt2phy(): cannot open /proc/self/pagemap: Permission denied
EAL: Cannot obtain physical addresses: Permission denied. Only vfio will
function.
EAL: Couldn't get fd on hugepage file
error allocating rte services array
EAL: FATAL: rte_service_init() failed
EAL: rte_service_init() failed
dumpcap: Can't get list of interfaces: Can't look for DPDK devices: A
service core failed to launch successfully

I modified the code to print the errorcode
TCL-sniffer3:~/build-pcap$ DPDK_CFG=--log-level=debug sudo dumpcap -D
EAL: Cannot obtain physical addresses: No such file or directory. Only vfio
will function.
error allocating rte services array
EAL: FATAL: rte_service_init() failed
EAL: rte_service_init() failed
dumpcap: Can't get list of interfaces: Can't look for DPDK devices: A
service core failed to launch successfully(err -8)

Regards
Anders

Den sön 10 feb. 2019 kl 22:21 skrev Guy Harris <notifications@github.com>:

**Comment by guyharris** [comment]
> I added the v18.11 packages to my LTS 18.04 Bionic by adding this line:
> deb [trusted=yes] http://ftp.se.debian.org/debian/ testing main contrib
> to /etc/apt/sourceslist it stil whines about the GPG Key but loads the package(s), Perhaps adding [trusted=yes] would work for travis too

That worked, although, unfortunately, the default Travis Ubuntu version is Trusty, [which is based on jessie](https://askubuntu.com/questions/445487/what-debian-version-are-the-different-ubuntu-versions-based-on), and [the jessie-backports version of DPDK is 16.11](https://packages.debian.org/jessie-backports/dpdk), and that's not recent enough to support building the DPDK support in libpcap.  The other Travis Ubuntu version is Xenial, [which is baed on stretch](https://askubuntu.com/questions/445487/what-debian-version-are-the-different-ubuntu-versions-based-on), and the [stretch-backports version of DPDK is 18.11](https://packages.debian.org/stretch-backports/dpdk), so I'll either go back to the matrix doing both Trusty and Xenial builds or just do Xenial builds.

**Comment by guyharris** [comment]
>     @TCL-sniffer3:~/build-pcap$ DPDK_CFG=--log-level=debug dumpcap -D
>     EAL: Detected 12 lcore(s)
>     EAL: Detected 1 NUMA nodes
>     EAL: Multi-process socket /run/user/1002/dpdk/rte/mp_socket
>     EAL: Probing VFIO support...
>     EAL: rte_mem_virt2phy(): cannot open /proc/self/pagemap: Permission denied
>     EAL: Cannot obtain physical addresses: Permission denied. Only vfio will function.
>     EAL: Couldn't get fd on hugepage file
>     error allocating rte services array
>     EAL: FATAL: rte_service_init() failed
>     EAL: rte_service_init() failed
>     dumpcap: Can't get list of interfaces: Can't look for DPDK devices: A service core failed to launch successfully

Unfortunately, "Couldn't get fd on hugepage file" just means that `get_seg_fd()` failed, which means either that `internal_config.in_memory` is true and `get_seg_memfd()` failed or that `internal_config.in_memory` is false and the file whose pathname was provided by `eal_get_hugefile_path()` couldn't be open for some reason.  It doesn't indicate whether that's because of a permissions problem or something else, even though `open()` *can* fail with EACCES, so there *should* be a way for `get_seg_fd()` to report *why* the failure occurred and its caller *should* determine why and return an indication of why the failure occurred.

**Pull request closed without merging**